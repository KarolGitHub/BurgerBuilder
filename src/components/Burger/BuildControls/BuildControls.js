import React from 'react';

import classes from './BuildControls.module.css';
import BuildControl from './BuildControl/BuildControl';
import PropTypes from 'prop-types';

const controls = [
    { label: 'Salad', type: 'salad' },
    { label: 'Cheese', type: 'cheese' },
    { label: 'Meat', type: 'meat' },
    { label: 'Bacon', type: 'bacon' },
    { label: 'Bread', type: 'bread' },
];

const buildControls = (props) => (
    <div className={classes.BuildControls}>
        <p>Total Price: <strong>{props.price.toFixed(2)}</strong></p>
        {controls.map((control) => <BuildControl
            key={control.label}
            label={control.label}
            added={() => props.ingredientAdded(control.type)}
            removed={() => props.ingredientRemoved(control.type)}
            disabled={props.disabled[control.type]}
        />)}
        <button
            className={classes.OrderButton}
            disabled={!props.purchaseable}
            onClick={props.purchased}
        >ORDER NOW
        </button>
    </div>
);

buildControls.propTypes = {
    price: PropTypes.number.isRequired,
    ingredientAdded: PropTypes.func.isRequired,
    ingredientRemoved: PropTypes.func.isRequired,
    disabled: PropTypes.object.isRequired,
    purchaseable: PropTypes.bool.isRequired,
    purchased: PropTypes.func.isRequired,
}
export default buildControls;
