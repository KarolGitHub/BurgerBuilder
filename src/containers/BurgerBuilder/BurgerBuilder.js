import React, { Component } from 'react';
import Aux from '../../Hoc/hoc/Auxiliary';
import { connect } from 'react-redux';

import Burger from '../../components/Burger/Burger';
import BuildControls from '../../components/Burger/BuildControls/BuildControls';
import Modal from '../../components/UI/Modal/Modal';
import OrderSummary from '../../components/Burger/OrderSummary/OrderSummary';
import axios from '../../axios-orders';
import Spinner from '../../components/UI/Spinner/Spinner';
import withErrorHandler from '../../Hoc/withErrorHandler/withErrorHandler';
import * as actions from '../../store/actions/index';
export class BurgerBuilder extends Component {

    state = {
        isModal: false,
        loading: false,
    }

    componentDidMount() {
        // console.log(this.props);
        this.props.onIngredientSet(this.props.building);
    }

    updatePurchaseState = (ingredients) => {
        const sum = Object.keys(ingredients)
            .map(igKey => ingredients[igKey])
            .reduce((sum, el) => {
                return sum + el;
            }, 0);
        return sum > 0;
    }

    modalHandler = (prevState) => {
        if (this.props.isAuth) {
            this.setState({ isModal: !prevState });
        }
        else {
            this.props.history.push('/login');
        }
    }

    purchaseContinueHandler = () => {

        this.props.history.push('/checkout');
        this.props.onInitPurchase();
    }

    render() {
        let disabledInfo = {
            ...this.props.ings,
            bread: this.props.bread
        };
        for (let key in disabledInfo) {
            disabledInfo[key] = disabledInfo[key] <= 0;
        }
        return (
            this.props.error ?
                null :
                this.props.ings ?
                    <Aux>
                        <Modal
                            isOpen={this.state.isModal}
                            isClosed={() => this.modalHandler(this.state.isModal)}>
                            {!this.state.loading ?
                                <OrderSummary
                                    ingredients={this.props.ings}
                                    cancel={() => this.modalHandler(this.state.isModal)}
                                    continue={this.purchaseContinueHandler}
                                    price={this.props.price} /> : <Spinner />}
                        </Modal>

                        <Burger ingredients={this.props.ings} bread={this.props.bread} />
                        <BuildControls
                            ingredientAdded={this.props.onIngredientAdded}
                            ingredientRemoved={this.props.onIngredientRemoved}
                            disabled={disabledInfo}
                            price={this.props.price}
                            purchaseable={this.updatePurchaseState(this.props.ings)}
                            purchased={() => this.modalHandler(this.state.isModal)}
                            isAuth={this.props.isAuth} />
                    </Aux> : <Spinner />
        );
    }
}

const mapStateToProps = state => {
    return {
        ings: state.burgerBuilder.ingredients,
        bread: state.burgerBuilder.bread,
        price: state.burgerBuilder.totalPrice,
        error: state.burgerBuilder.error,
        building: state.burgerBuilder.building,
        isAuth: state.auth.token !== null,
    };
}

const mapDispatchtoProps = dispatch => {
    return {
        onIngredientAdded: (ing) => dispatch(actions.addIng(ing)),
        onIngredientRemoved: (ing) => dispatch(actions.remIng(ing)),
        onIngredientSet: (building) => dispatch(actions.setIng(building)),
        onInitPurchase: () => dispatch(actions.purchaseInit())
    }
}

export default connect(mapStateToProps, mapDispatchtoProps)(withErrorHandler(BurgerBuilder, axios));