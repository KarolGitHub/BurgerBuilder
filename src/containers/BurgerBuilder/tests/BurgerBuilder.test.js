import React from "react";
import { shallow } from "enzyme";

import BurgerBuilder from "../BurgerBuilder";
import BuildControls from "../../../components/Burger/BuildControls/BuildControls";

describe("<BurgerBuilder />", () => {
  describe("Unit tests", () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<BurgerBuilder onIngredientSet={() => {}} />);
    });

    it("should render <BuildControls /> when receiving ingredients", () => {
      wrapper.setProps({ ings: { salad: 0 } });
      expect(wrapper.find(BuildControls)).toBe({});
    });
  });
});
