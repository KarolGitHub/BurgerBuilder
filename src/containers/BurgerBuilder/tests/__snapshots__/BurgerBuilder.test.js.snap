// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BurgerBuilder /> Unit tests snapshot match 1`] = `undefined`;

exports[`<NavigationItems /> Unit tests snapshot match 1`] = `undefined`;
